rust/src/measurements/gumbel_max/make_report_noisy_max_gumbel.tex

```diff
@@ -1,253 +1,248 @@
+\begin{itemize} 
+    \item \texttt{TIA} (atomic input type) is a type with trait \rustdoc{traits/trait}{Number}
+    \item \texttt{f64} is a type with trait \rustdoc{traits/trait}{DistanceConstant}\texttt{<TIA>}
+    \item \texttt{FBig} is a type that supports fallible exact casting from \texttt{TIA} and \texttt{f64}
+\end{itemize} 
+
+\textbf{Human-verified}
+None
+ 
+\subsection*{Function} 
+\label{sec:python-pseudocode} 
+\lstinputlisting[language=Python,firstline=2]{./pseudocode/make_report_noisy_max_gumbel.py} 
+ 
+\subsection*{Postcondition} 
+ 
+\validMeasurement{\texttt{input\_domain, input\_metric, scale, optimize, TIA}}{\texttt{make\_report\_noisy\_max\_gumbel}}  
+ 
+\section{Proof} 
+\subsection{Privacy Guarantee} 
+ 
+To ensure that the Gumbel sample is valid, the \texttt{input\_domain} is required to be non-null. 
+The scale is also required to be positive. 
+ 
+\begin{lemma} 
+    \label{function-prob} 
+    By the definition of $\function$ in the pseudocode, for any $x$ in \texttt{input\_domain}, \\ 
+    $\Pr[\function(x) = i] = \Pr[\mathrm{argmax}_k (u_k / \tau - \ln(-\ln(U_k))) = i]$. 
+\end{lemma} 
+ 
+\begin{proof} 
+    For each score $s_k$, \function\ samples a Gumbel random variable centered at $sign \cdot s_k / \tau$. 
+    The choice of sign does not affect the privacy guarantee, so we omit it from further analysis. 
+    Sampling from a Gumbel distribution is equivalent to adding a draw from $-\ln(-\ln(U_k))$, where $U_k \sim Uniform(0, 1)$. 
+    The algorithm only returns the index of the maximum Gumbel random variable, 
+    therefore the probability of returning $i$ is the probability that the $i^{th}$ Gumbel random variable is the maximum. 
+\end{proof} 
+ 
+\begin{lemma} 
+    \label{priv-inequality} 
+    Assume $x$, $x'$ in \texttt{input\_domain}. Then 
+    $\ln\left(\frac{\sum_{i} \exp(\frac{\epsilon x'_i}{\Delta})}{\sum_{i} \exp(\frac{\epsilon x_i}{\Delta})}\right) \le \frac{\epsilon \max_j (x'_j - x_j)}{\Delta}$. 
+\end{lemma} 
+ 
+\begin{proof} 
+\begin{align*} 
+    \ln\left(\frac{\sum_{i} \exp(\frac{\epsilon x'_i}{\Delta})}{\sum_{i} \exp(\frac{\epsilon x_i}{\Delta})}\right) 
+    &= \ln\left(\frac{\sum_{i} \exp(\frac{\epsilon (x'_i - x_i + x_i)}{\Delta})}{\sum_{i} \exp(\frac{\epsilon x_i}{\Delta})}\right) \\ 
+    &= \ln\left(\frac{\sum_{i} \exp(\frac{\epsilon (x'_i - x_i)}{\Delta})\exp(\frac{\epsilon (x_i)}{\Delta})}{\sum_{i} \exp(\frac{\epsilon x_i}{\Delta})}\right) \\ 
+    &\le \ln\left(\frac{\exp(\frac{\epsilon \max_j(x'_j - x_j)}{\Delta}) \sum_{i} \exp(\frac{\epsilon x_i}{\Delta})}{\sum_{i} \exp(\frac{\epsilon x_i}{\Delta})}\right) \\ 
+    &= \frac{\epsilon \max_j(x'_j - x_j)}{\Delta} 
+\end{align*} 
+\end{proof} 
+ 
+\label{privacy-guarantee} 
+Assume $x$, $x'$ in \texttt{input\_domain} are \din-close under \rustdoc{metrics/struct}{LInfDistance} and $\texttt{privacy\_map}(\din) \le \dout$. 
+Let the output random variables be denoted $Y \sim \function(x)$ and $Y' \sim \function(x')$. 
```

Comment:
"Then we have the following." Or something like that at the end of this sentence.